### PR TITLE
Large Movement Code Optimization

### PR DESCRIPTION
**Pull request by Kylerace**
## About The Pull Request
ive been working on this on and off for a few months and its finally to the point i can test it. tries to optimize the movement process where feasible. unfortunately i mostly forgot what the benchmark results are, guess we'll see live. most of movements cost is player mobs moving around so a lot of what i did is related to things hooked onto player movement. 
the biggest improvement is likely in footsteps and playsound() in general. I was able to use the spatial grid to 
1. make the search for other player mobs as part of playsound() more efficient by searching through the grid, if i remember correctly it beat the previous method (iterating through all the players on the z level) at around 50 players, which is good because we save on cpu during highpop

2. make playsound() not need to search for other players for footsteps the majority of the time using a cached list of player mobs in adjacent spatial grid cells. when a player moves from one cell to another the game searches for all other players in the 8 spatial grid cells adjacent to the one the player is located in (as well as the center one) and adds a reference to them in `closeby_client_mobs` and a reference to us in theirs, and also unsets all of the previously linked clients that are no longer adjacent. since it updates the link both ways and updates every time a player mob changes grid cells, this should guarantee that every player mob knows about every other player mob within at least `SPATIAL_CELL_SIZE + 1` tiles, and thus we can use this list for sounds emanating from the player mob as long as range is less than that amount, so we use it for footsteps. this is good because we only need to update this data when any given player moves spatial grid cells which are 17x17 turfs in area, but footsteps happen every couple of turfs, which would normally require finding players in range for each 

for a visual representation of the second point, consider this nonsensical diagram
Screenshot_1813
every dot is a player mob, and the connections represent a reference in each mobs `closeby_client_mobs` list. when any player moves spatial grid cells each players list is updated.
this shouldnt be a massive improvement of footstep cost, but should be nice.

also, the fov effects no longer have to call get_hearers_in_view() when called as part of the footstep process, they just use the list of client mobs that playsound() returned. this means that footsteps can cause field of view effects from further than 5 tiles away but ill take that

another thing thats done is some optimizations to the turf/Entered() proc and proc calls originating from there, since that usually takes between 1/20th to 1/10th of the cost of movable/Move() varying quite a bit between rounds. 

frozen objects are now a simple element as opposed to something that needs to hook onto turf/Entered(), and also turf/Entered() is totally removed since nothing uses it anymore.

z movement being blocked by an object on the turf now uses a signal system with connect_loc as opposed to iterating through contents to look for objects that block z movement. this might actually be slower overall due to the added overhead of signals and the relative low number of z movement blocking objects on tiles where z movement happens, so i might remove this in the end.

also adds active_storage objects to important_recursive_contents. which means that objects which have a storage component that a player has opened (ie theyre looking at its contents) the object that storage component is attached to will be in important_recursive_contents. this is because if a player has a currently active storage component, mob/living/Move() will check if they can still access that component and will close it if they cant. the problem is that this check involves getting all nested player contents in CanReach() and doing a lot of useless iterations on top of that and active_storage can point to something in the players direct contents (like their backpack). so to avoid doing that expensive check every movement for a significant portion of players it first checks if the active storage object is in the players important_recursive_contents and if it is then it doesnt have to do that check.

## Why It's Good For The Game
movement across practically every round is expensive. SSinput can easily take ~10% of the tick every tick at 60+ pop and since movement code is broad enough theres a decent sized target to cut down on that cost.

## Changelog
🆑 
balance: for performance reasons (believe it or not), field of view effects from people walking around outside of your field of view can now happen further than 5 tiles away.
/🆑

### CODE REVIEW

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 1]
```diff
@@ -113,7 +113,7 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 		var/cached_gases = gases
 		TOTAL_MOLES(cached_gases, .)
 		. *= R_IDEAL_GAS_EQUATION * temperature / volume
-		return
+		return .
```
Comment (Ghilker) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm:116]: Unrelated movement change pings me
image

Comment (Kylerace) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:116]: :trollface:

**Comment by Kylerace** [comment]
reee ill fix it in post

File: code/game/atoms.dm [hunk 2]
```diff
@@ -339,20 +339,26 @@
...
 	if(mover.movement_type & PHASING)
 		return TRUE
-	. = CanAllowThrough(mover, border_dir)
+	. = CanAllowThrough(mover, border_dir) //very complicated and is where most of the mojo is. many overrides
+	//however this is the only big thing i need to check for refactoring
+
 	// This is cheaper than calling the proc every time since most things dont override CanPassThrough
 	if(!mover.generic_canpass)
-		return mover.CanPassThrough(src, REVERSE_DIR(border_dir), .)
+		return mover.CanPassThrough(src, REVERSE_DIR(border_dir), .)//this can be changed with registering to a signal sent in turf/Enter
```
Comment (Kylerace) [inline code/game/atoms.dm:347]:
```suggestion
		return mover.CanPassThrough(src, REVERSE_DIR(border_dir), .)
```

**Comment by optimumtact** [comment]
`closeby_client_mobs` more like `free_hard_dels_here`

**Comment by LemonInTheDark** [comment]
Why is this unprofilable exactly?

**Comment by Kylerace** [comment]
more that any profile cant represent the truth of how this affects movement cost since its affected by a lot of factors averaging out, only live could really do that since cost behavior is mainly determined by player action. i could make a general and maybe 1-2 specific benchmarks if you want

File: code/datums/elements/frozen.dm [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	RegisterSignal(target, COMSIG_MOVABLE_POST_THROW, .proc/shatter_on_throw)
+	RegisterSignal(target, COMSIG_OBJ_UNFREEZE, .proc/on_unfrozen)
+
+/datum/element/frozen/Detach(datum/source, ...)
+	var/obj/obj_source = source
+	REMOVE_TRAIT(obj_source, TRAIT_FROZEN, TRAIT_GENERIC)
+	obj_source.name = replacetext(obj_source.name, "frozen ", "")
+	obj_source.remove_atom_colour(TEMPORARY_COLOUR_PRIORITY, GLOB.freon_color_matrix)
+	obj_source.alpha += 25
+	. = ..()
```
Comment (Watermelon914) [inline code/datums/elements/frozen.dm:29]:
```suggestion
	return ..()
```

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/world.dm [hunk 4]
```diff
@@ -19,6 +19,7 @@
 	hub_password = "kMZy3U5jJHSiBQjr"
 	name = "/tg/ Station 13"
 	fps = 20
+	loop_checks = FALSE
 #ifdef FIND_REF_NO_CHECK_TICK
 	loop_checks = FALSE
 #endif
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/world.dm:25]: why are you setting loop checks to false? that's an important fallback

[the remark this answers is not in the document]
Comment (Kylerace) [reply code/world.dm:25]: because i was benchmarking and forgot to put it back

File: code/modules/vehicles/vehicle_actions.dm [hunk 5]
```diff
@@ -293,7 +293,7 @@
 	. = ..()
 	if(.)
 		var/obj/vehicle/ridden/scooter/skateboard/vehicle = vehicle_target
-		vehicle.obj_flags |= BLOCK_Z_OUT_DOWN
+		//vehicle.obj_flags |= BLOCK_Z_OUT_DOWN
```
[the remark this answers is not in the document]
Comment (Kylerace) [reply code/modules/vehicles/vehicle_actions.dm]: i didnt like my refactor of that, so i reverted it and it works like it does on master again

File: code/modules/mob/living/living.dm [hunk 6]
```diff
@@ -64,105 +64,98 @@
...
+		if(our_disease.spread_flags & DISEASE_SPREAD_CONTACT_SKIN)
+			bumping_mob.ContactContractDisease(our_disease)
+
+	for(var/datum/disease/their_disease as anything in bumping_mob.diseases)
+		if(their_disease.spread_flags & DISEASE_SPREAD_CONTACT_SKIN)
+			ContactContractDisease(their_disease)
+
+	//Should stop you pushing a restrained person out of the way
+	if(bumping_mob.pulledby && bumping_mob.pulledby != src && HAS_TRAIT(bumping_mob, TRAIT_RESTRAINED))
+		if(!(world.time % 5)) //?????????
```
[the remark this answers is not in the document]
Comment (Kylerace) [reply code/modules/mob/living/living.dm:105]: that already existed i just moved it a bit, i couldnt figure out why it was done like that.

**Comment by LemonInTheDark** [comment]
This been testmerged?

**Comment by Kylerace** [comment]
taking a break to focus on school
